Update usaco-616.mdx + add java/cpp code (#1866)
cpinitiative/usaco-guide: solutions/usaco-616.mdx

id: usaco-616
source: USACO Bronze 2016 February
title: Circular Barn
author: Varun Ragunath
---

[Official Analysis](http://www.usaco.org/current/data/sol_cbarn_bronze_feb16.html)

> ## Implementation

Reviewer: add time complexity

**Time Complexity:** $\mathcal{O}(N^2)$

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
	freopen("cbarn.in", "r", stdin);
	freopen("cbarn.out", "w", stdout);
	cin.sync_with_stdio(0); cin.tie(0);

	int n;
	cin >> n;

	vector<int> cows(n);

	for(int i = 0; i < n; i++) {
		cin >> cows[i];
	}

	// run a simple brute force algo, checking each possible barn we can open

	int ret = (int)1e9; // stores the current answer
	// initially set to an abnormally large number

	for(int open_barn = 0; open_barn < n; open_barn++) {
		int curdist = 0; 
		// stores the current distance if barn i was the opened barn
		for(int pos = 0; pos < n; pos++) {
			// the current barn we are processing is open_barn + pos
			// however, since it is a circular array 
			// the position may wrap around
			// index = open_barn + pos
			curdist += pos * cows[(open_barn + pos) % n];
		}
		ret = min(ret, curdist);
	}

	cout << ret << '\n';

> 	return 0;

Reviewer 2: ```suggestion
```

## Implementation

```cpp

> #include<bits/stdc++.h>

Reviewer 2: ```suggestion
#include <bits/stdc++.h>
```

using namespace std;

int main() {
	freopen("cbarn.in", "r", stdin);
	freopen("cbarn.out", "w", stdout);
	cin.sync_with_stdio(0); cin.tie(0);

	int n;
	cin >> n;

	vector<int> cows(n);

	for(int i = 0; i < n; i++) {
		cin >> cows[i];
	}

> 	// run a simple brute force algo, checking each possible barn we can open
> 
> 	int ret = (int)1e9; // stores the current answer
> 	// initially set to an abnormally large number
> 

Reviewer 2: ```suggestion
	// run a simple brute force algo that checks each possible barn
	int ret = INT32_MAX;
```

	for(int open_barn = 0; open_barn < n; open_barn++) {
		int curdist = 0; 
		// stores the current distance if barn i was the opened barn
		for(int pos = 0; pos < n; pos++) {
			// the current barn we are processing is open_barn + pos
			// however, since it is a circular array 
			// the position may wrap around
			// index = open_barn + pos
			curdist += pos * cows[(open_barn + pos) % n];
		}
		ret = min(ret, curdist);
	}

  cout << ret << '\n';

> 
>   return 0;

Reviewer 2: ```suggestion
```

> ## Implementation

Reviewer: you can add your name to the author list at the top

```cpp
#include<bits/stdc++.h>
using namespace std;

> int main() {

Reviewer: use tabs instead of spaces to indent

  freopen("cbarn.in", "r", stdin);
  freopen("cbarn.out", "w", stdout);
  cin.sync_with_stdio(0); cin.tie(0);

>   // input
> 
>   // have the number of cows and desired positions of each cow
> 

Reviewer 2: ```suggestion
```

  int n;
  cin >> n;

  vector<int> cows(n);

  for(int i = 0; i < n; i++) {
    cin >> cows[i];
  }

  // simple brute force

  // check each possible barn we can open

> 

Reviewer 2: ```suggestion
  // run a simple brute force algo, checking each possible barn we can open

>   int ret = (int)1e9; // stores the current answer
>   // initially set to an abnormally large number
> 

Reviewer 2: ```suggestion
  int ret = INT32_MAX;
```

  for(int open_barn = 0; open_barn < n; open_barn++) {
    int curdist = 0; // stores the current distance if barn i was the opened barn
    for(int pos = 0; pos < n; pos++) {

>       // the current barn we are processing is open_barn + pos

Reviewer: use multiline comments?

      // however, since it is a circular array 
      // the position may wrap around

>       // index = (open_barn + pos)

Reviewer: ```suggestion
      // index = open_barn + pos
```

>       curdist += pos * cows[(open_barn + pos)%n];

Reviewer: ```suggestion
      curdist += pos * cows[(open_barn + pos) % n];
```

    }

>     // update the answer

Reviewer 2: ```suggestion
```

    ret = min(ret, curdist);
  }

> 

Reviewer 2: ```suggestion
```

  cout << ret << '\n';

  return 0;
}

```

</CPPSection>
	
<JavaSection>
	
## Implementation	
	
```java
import java.io.*;
import java.util.*;

> public class cbarn {

Reviewer 2: ```suggestion

public class CBarn {
```

	public static void main(String[] args) throws IOException {

> 		// initialize file I/O

Reviewer 2: ```suggestion
```

		BufferedReader br = new BufferedReader(new FileReader("cbarn.in"));
		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("cbarn.out")));
		// read in N
		int n = Integer.parseInt(br.readLine());
		int[] cows = new int[n];
		// read in how many cows need to be in each room
		for(int i = 0; i < n; i++) {
			cows[i] = Integer.parseInt(br.readLine());
		}
		// the answer cannot exceed N * N * 100, since there are at most 100N cows, each of which can move at most N
		int ans = n * n * 100;
		for(int unlock = 0; unlock < n; unlock++) {
			// assume we unlock the door at index "unlock", compute the distance all cows travel
			int currentDistance = 0;
			for(int offset = 0; offset < n; offset++) {
				// count how many cows have to walk a distance of "offset"

> 				currentDistance += offset * cows[(unlock+offset)%n];

Reviewer: ```suggestion
				currentDistance += offset * cows[(unlock + offset) % n];
```

Reviewer: bump https://github.com/cpinitiative/usaco-guide/pull/1866/files#r663159957

			}
			// update the answer
			if(currentDistance < ans) {
				ans = currentDistance;
			}
		}
		// print the answer
		pw.println(ans);
		// close output stream
		pw.close();

> 		br.close();

Reviewer 2: lines 78, 81, 94, 99, and 101 are unnecessary comments (they shouldn't explain what you're doing, but explain _why_ you're doing what you're doing!)

```cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
	freopen("cbarn.in", "r", stdin);
	freopen("cbarn.out", "w", stdout);
	cin.sync_with_stdio(0); cin.tie(0);

	int n;
	cin >> n;

	vector<int> cows(n);

	for(int i = 0; i < n; i++) {
		cin >> cows[i];
	}

	// run a simple brute force algo, checking each possible barn we can open

	int ret = (int)1e9; // stores the current answer
	// initially set to an abnormally large number

	for(int open_barn = 0; open_barn < n; open_barn++) {
		int curdist = 0; 
		// stores the current distance if barn i was the opened barn
		for(int pos = 0; pos < n; pos++) {
			// the current barn we are processing is open_barn + pos
			// however, since it is a circular array 
			// the position may wrap around
			// index = open_barn + pos
			curdist += pos * cows[(open_barn + pos) % n];
		}
		ret = min(ret, curdist);
	}

  cout << ret << '\n';

>   return 0;

Reviewer: indentation here is weird

}

```

</CPPSection>
	
<JavaSection>
	

> ## Implementation	

Reviewer: move ## Implementation outside languagesection and remove ## Implementation from cppsection

	
```java
import java.io.*;
import java.util.*;
public class cbarn {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new FileReader("cbarn.in"));

> 		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("cbarn.out")));

Reviewer: don't need these since you're using kattio, right?

	
		cbarn.Kattio io = new cbarn.Kattio("cbarn");
		
		int n = io.nextInt();
		int[] cows = new int[n];
		for(int i = 0; i < n; i++) {
			cows[i] = io.nextInt();
		}
		// the answer cannot exceed N * N * 100, 
		// since there are at most 100N cows, each of which can move at most N
		int ans = n * n * 100;
		for(int unlock = 0; unlock < n; unlock++) {
			// assume we unlock the door at index "unlock", compute the distance all cows travel
			int currentDistance = 0;
			for(int offset = 0; offset < n; offset++) {
				// count how many cows have to walk a distance of "offset"
				currentDistance += offset * cows[(unlock+offset)%n];
			}
			ans = Math.min(ans, currentDistance);
		}
		io.println(ans);
		io.close();
	}
	// https://usaco.guide/general/io?lang=java#io-template

> 	static class Kattio extends PrintWriter {

Reviewer: replace the entire kattio with `CodeSnip{Kattio}`

	for(int i = 0; i < n; i++) {
		cin >> cows[i];
	}

	// run a simple brute force algo, checking each possible barn we can open

	int ret = (int)1e9; // stores the current answer
	// initially set to an abnormally large number

	for(int open_barn = 0; open_barn < n; open_barn++) {
		int curdist = 0; 
		// stores the current distance if barn i was the opened barn
		for(int pos = 0; pos < n; pos++) {
			// the current barn we are processing is open_barn + pos
			// however, since it is a circular array 
			// the position may wrap around
			// index = open_barn + pos
			curdist += pos * cows[(open_barn + pos) % n];
		}
		ret = min(ret, curdist);
	}

  cout << ret << '\n';

  return 0;
}

```

</CPPSection>
	
<JavaSection>
	
## Implementation	
	
```java
import java.io.*;
import java.util.*;
public class cbarn {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new FileReader("cbarn.in"));
		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("cbarn.out")));
		
		int n = Integer.parseInt(br.readLine());
		int[] cows = new int[n];
		for(int i = 0; i < n; i++) {
			cows[i] = Integer.parseInt(br.readLine());
		}

> 		// the answer cannot exceed N * N * 100, since there are at most 100N cows, each of which can move at most N

Reviewer 3: this line is over the limit

		int ans = n * n * 100;
		for(int unlock = 0; unlock < n; unlock++) {
			// assume we unlock the door at index "unlock", compute the distance all cows travel
			int currentDistance = 0;
			for(int offset = 0; offset < n; offset++) {
				// count how many cows have to walk a distance of "offset"
				currentDistance += offset * cows[(unlock+offset)%n];
			}

> 			if(currentDistance < ans) {

Reviewer 3: can be simplified using minimum

				ans = currentDistance;
			}
		}
		pw.println(ans);

> 		pw.close();

Reviewer 3: i dont think kattio is require but should look nicer its just one close
